vpr/src/place/analytic_placer.h

```diff
@@ -0,0 +1,249 @@
+
+    // @see initial_placement.cpp load_legal_locations()
+    std::vector<std::vector<int>> num_legal_pos;
+    std::vector<std::vector<std::vector<t_pl_loc>>> legal_pos;
+
+    // marker for blks not solved in current iteration
+    int32_t dont_solve = std::numeric_limits<int32_t>::max();
+
+    // row number in the system of linear equations for each block
+    std::unordered_map<ClusterBlockId, int32_t> blk_row_num;
+
+    // Encapsulates 3 types of locations for each logic block
+    struct BlockLocation {
+        t_pl_loc loc;       // real, up-to-date location of the logic block
+        t_pl_loc legal_loc; // legalized location, used to create psudo connections
+        double rawx, rawy;  // raw location storing float result from matrix solver
+    };
+
+    // Lookup from blockID to block location
+    std::unordered_map<ClusterBlockId, BlockLocation> blk_locs;
+
+    // The set of blks actually placed.
+    // Excludes children of macros.
+    std::vector<ClusterBlockId> place_blks;
+
+    // blocks being solved in the current equation
+    // subset of all blocks being placed
+    std::vector<ClusterBlockId> solve_blks;
+
+    /*
+     * number of blocks on each tile
+     * used to print a simple placement graph in print_place
+     */
+    std::vector<std::vector<int>> overlap;
+
+    /*
+     * Prints the location of each block, and a simple drawing of FPGA fabric, showing num of blocks on each tile
+     * Very useful for debugging
+     * See implementation for usage
+     */
+    void print_place(const char* place_file);
+
+    //build fast lookup of tiles/subtiles by tile, x, y, subtiles
+    void build_fast_tiles();
+
+    // build num_legal_pos, legal_pos like in initial_placement.cpp
+    void build_legal_locations();
+
+    // build blk_locs based on initial placement;
+    // build and place_blks for blocks that needs to be placed;
+    void init();
+
+    wirelen_t get_net_hpwl(ClusterNetId net_id);
+
+    wirelen_t total_hpwl();
+
+    // Setup the blocks of type blkTypes to be solved
+    // Returns the number of rows (number of blks to solve)
+    int setup_solve_blks(t_logical_block_type_ptr blkTypes);
+
+    // Update the location of all members of all macros based on location of macro_head
+    // Since only macro_head is solved (connections to macro members are also taken into account
+    // when formulating the matrix equations)
+    void update_macros();
+
+    // Build and solve in one direction
+    // Solved solutions are written back to block_locs in rawx, rawy
+    void build_solve_direction(bool yaxis, int iter, int build_solve_iter);
+
+    // Build the system of equations for either X or Y
+    // When iter > 1, psudo-conenctions are formed, the strength is determined by alpha and iter
+    void build_equations(EquationSystem<double>& es, bool yaxis, int iter = -1);
+
+    // Solve the system of equations
+    void solve_equations(EquationSystem<double>& es, bool yaxis);
```

Comment:
explain yaxis.  What does this update (must be a class member variable, as it appears to have no output from the prototype).